task1Daniv/stringlist.cpp

```diff
@@ -0,0 +1,250 @@
+#include "stringlist.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+const int dataIndex = 0;
+const int nextNodeIndex = 1;
+
+List initialization()
+{
+    List temp = static_cast<char**>(malloc(sizeof(List) * 2));
+
+    if(temp != nullptr)
+    {
+        temp[dataIndex] = nullptr;
+        temp[nextNodeIndex] = nullptr;
+    }
+    return  temp;
+}
+
+void push(List &list, const String &string)
+{
+    if(list == nullptr)
+    {
+        return;
+    }
+    String textStr = static_cast<String>(malloc(sizeof(char) * strlen(string) + 1));
+    strcpy(textStr, string);
+
+    if(list[dataIndex] == nullptr)
+    {
+        list[dataIndex] = textStr;
+    }
+    else
+    {
+        auto temp = list;
+        while(list[nextNodeIndex] != nullptr)
+        {
+            list = reinterpret_cast<List>(list[nextNodeIndex]);
+        }
+        List node = static_cast<List>(malloc(sizeof(char*) * 2));
+        node[dataIndex] = textStr;
+        node[nextNodeIndex] = nullptr;
+        list[nextNodeIndex] = reinterpret_cast<String>(node);
+        list = temp;
+    }
+}
+
+void print(const List &list)
+{
+    if(list == nullptr)
+    {
+        return;
+    }
+    auto temp = list;
+    for(int i = 0; i < size(list); ++i)
+    {
+        printf("%s\n",temp[dataIndex]);
+        temp = reinterpret_cast<List>(temp[nextNodeIndex]);
+
+    }
+}
+
+int size(const List &list)
+{
+    if(list == nullptr)
+    {
+        return -1;
+    }
+
+    int totalSize = 0;
+    auto temp = list;
+    while (temp != nullptr && temp[dataIndex] != nullptr)
+    {
+        totalSize++;
+        temp = reinterpret_cast<List>(temp[nextNodeIndex]);
+    }
+    return totalSize;
+}
+
+int index(const List &list, const String &string)
```

Comment:
You return **-1** etc. as signal combinations. This is bad practice since only signed types can have minus sign values. E.g. in case of interface changes, you have to also change the behavior of your program. The better practice here is to use the index one past the end ;)

Reply:
At all try not to use hardcoded values. Instead, move that values to the constants or constant expressions. Reader ought to understand what is the meaning of the value, you use.